src/util/translation.h

```diff
@@ -6,10 +6,43 @@
 #define BITCOIN_UTIL_TRANSLATION_H
 
 #include <tinyformat.h>
+#include <util/string.h>
 
+#include <cassert>
 #include <functional>
 #include <string>
 
+/** Translate a message to the native language of the user. */
+const extern std::function<std::string(const char*)> G_TRANSLATION_FUN;
+
+namespace util {
+/**
+ * Translation function.
+ * If no translation function is set, simply return the input.
+ */
+inline std::string translate(const char* lit)
+{
+    return G_TRANSLATION_FUN ? G_TRANSLATION_FUN(lit) : lit;
+}
+/** Type to denote whether an orginal string literal is translatable */
+template <bool translatable = true>
+struct Original {
+    const char* const lit;
+    consteval Original(const char* str) : lit{str} { assert(lit); }
+    std::string translate() const { return translatable ? util::translate(lit) : lit; }
+};
+template <unsigned num_params>
+struct bilingual_fmt {
+    const ConstevalFormatString<num_params> original;
+    const bool translatable;
+    template <bool translatable>
+    consteval bilingual_fmt(Original<translatable> o) : original{o.lit}, translatable{translatable}
+    {
+    }
+    std::string translate() { return translate(original.fmt); }
```

Comment:
re: https://github.com/bitcoin/bitcoin/pull/31061#discussion_r1793744937

This is a good catch, but it seems better to just drop the template method instead of trying to fix it up and use it. It just seems to add complexity and it doesn't seem like there is another use-case for it.